server/container/util/src/main/java/org/apache/james/util/StreamUtils.java

```diff
@@ -54,4 +59,42 @@
     public static <T> Stream<T> flatten(Stream<T>... streams) {
         return flatten(Arrays.stream(streams));
     }
+
+    public static <T> Stream<T> unfold(T seed, Function<T, Optional<T>> generator) {
+        return StreamSupport.stream(new UnfoldSpliterator(seed, generator), !PARALLEL);
+    }
+
+
+    private static class UnfoldSpliterator<T> implements Spliterator<T> {
+
+        private Optional<T> current;
+        private final Function<T, Optional<T>> generator;
+
+        private UnfoldSpliterator(T seed, Function<T, Optional<T>> generator) {
+            this.current = Optional.of(seed);
+            this.generator = generator;
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            current.ifPresent(action);
+            current = current.flatMap(generator);
+            return current.isPresent();
+        }
+
+        @Override
+        public Spliterator<T> trySplit() {
+            return null;
+        }
+
+        @Override
+        public long estimateSize() {
+            return Long.MAX_VALUE;
+        }
+
+        @Override
+        public int characteristics() {
+            return Spliterator.IMMUTABLE & Spliterator.NONNULL & Spliterator.ORDERED;
```

Comment:
Can you explain me why this spliterator is mutable while it maintain a state?

Reply:
It's immutable as declared here, so I do not understand

Reply:
How about the current field reallocated in tryAdvance?

Reply:
A spliterator is defined as immutable if the source is immutable, not the current value, cf. https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html#IMMUTABLE

So the question is can we consider a Java function as an immutable source? :trollface: